src/Microsoft.Azure.WebJobs.Host/Singleton/BlobLeaseDistributedLockManager.cs

```diff
@@ -0,0 +1,460 @@
+            string account,
+            string lockId,
+            string lockOwnerId,
+            TimeSpan lockPeriod,
+            CancellationToken cancellationToken)
+        {
+            IStorageBlobDirectory lockDirectory = GetLockDirectory(account);
+            IStorageBlockBlob lockBlob = lockDirectory.GetBlockBlobReference(lockId);
+            string leaseId = await TryAcquireLeaseAsync(lockBlob, lockPeriod, cancellationToken);
+
+            if (string.IsNullOrEmpty(leaseId))
+            {
+                return null;
+            }
+
+            if (!string.IsNullOrEmpty(lockOwnerId))
+            {
+                await WriteLeaseBlobMetadata(lockBlob, leaseId, lockOwnerId, cancellationToken);
+            }
+
+            SingletonLockHandle lockHandle = new SingletonLockHandle
+            {
+                LeaseId = leaseId,
+                LockId = lockId,
+                Blob = lockBlob
+            };
+            lockHandle.LeaseRenewalTimer = CreateLeaseRenewalTimer(lockBlob, leaseId, lockId, lockPeriod, _exceptionHandler, lockHandle);
+            
+            // start the renewal timer, which ensures that we maintain our lease until
+            // the lock is released
+            lockHandle.LeaseRenewalTimer.Start();
+
+            return lockHandle;
+        }
+
+        internal IStorageBlobDirectory GetLockDirectory(string accountName)
+        {
+            if (string.IsNullOrEmpty(accountName))
+            {
+                accountName = ConnectionStringNames.Storage;
+            }
+
+            IStorageBlobDirectory storageDirectory = null;
+            if (!_lockDirectoryMap.TryGetValue(accountName, out storageDirectory))
+            {
+                Task<IStorageAccount> task = _accountProvider.GetStorageAccountAsync(accountName, CancellationToken.None);
+                IStorageAccount storageAccount = task.Result;
+                // singleton requires block blobs, cannot be premium
+                storageAccount.AssertTypeOneOf(StorageAccountType.GeneralPurpose, StorageAccountType.BlobOnly);
+                IStorageBlobClient blobClient = storageAccount.CreateBlobClient();
+                storageDirectory = blobClient.GetContainerReference(HostContainerNames.Hosts)
+                                       .GetDirectoryReference(HostDirectoryNames.SingletonLocks);
+                _lockDirectoryMap[accountName] = storageDirectory;
+            }
+
+            return storageDirectory;
+        }
+
+        private static async Task<string> TryAcquireLeaseAsync(IStorageBlockBlob blob, TimeSpan leasePeriod, CancellationToken cancellationToken)
+        {
+            bool blobDoesNotExist = false;
+            try
+            {
+                // Optimistically try to acquire the lease. The blob may not yet
+                // exist. If it doesn't we handle the 404, create it, and retry below
+                return await blob.AcquireLeaseAsync(leasePeriod, null, cancellationToken);
```

Comment:
Does TryLockAsync.LockId solve that? 

---
In reply to: [120949833](https://github.com/Azure/azure-webjobs-sdk/pull/1184#discussion_r120949833) [](ancestors = 120949833)